SparseGrids/tsgGridLocalPolynomial.cpp

```diff
@@ -846,6 +872,48 @@ double GridLocalPolynomial::evalBasisSupported(const int point[], const double x
     return f;
 }
 
+std::vector<double> GridLocalPolynomial::diffBasisRaw(const int point[], const double x[]) const{
+    std::vector<double> df(num_dimensions, 0.0);
+    std::vector<double> basis_values(num_dimensions);
+    int num_zero_values = 0;
+    double nonzero_values_prod = 1.0;
+    int idx_zero_value;
+    bool supported_at_j;
+    for(int j=1; j<num_dimensions; j++) {
+        basis_values[j] = rule->evalSupport(point[j], x[j], supported_at_j);
+        if (std::fabs(basis_values[j]) < Maths::num_tol or !supported_at_j) {
+            num_zero_values++;
+            idx_zero_value = j;
+        } else {
+            nonzero_values_prod *= basis_values[j];
+        }
+    }
+    if (num_zero_values == 0)
+        for (int j=0; j<num_dimensions; j++)
+            df[j] += nonzero_values_prod / basis_values[j] * rule->diffRaw(point[j], x[j]);
+    else if (num_zero_values == 1)
+        df[idx_zero_value] += nonzero_values_prod * rule->diffRaw(point[idx_zero_value], x[idx_zero_value]);
+    return df;
+}
+std::vector<double> GridLocalPolynomial::diffBasisSupported(const int point[], const double x[], bool &isSupported) const{
+    isSupported = false;
+    std::vector<double> diff_values(num_dimensions, 1.0);
+    bool isDimSupported = false;
+    for(int k=0; k<num_dimensions; k++) {
+        for(int j=0; j<k; j++) {
+            diff_values[k] *= rule->evalSupport(point[j], x[j], isDimSupported);
```

Comment:
As you can see in the implementation of `evalSuppor()` the method is not trivial. The way you have ordered the for-loops, for different values of `k` you will have multiple calls to `evalSupport()` with identical values of `j`.

Make the outer loop over `j`, call `evalSuppor()` once, then have the two `k` loops to multiply the values that you need. You can compute the `rule->diffSupport()` either inside or in a separate loop.

Reply:
Changed. The calls to `rule->diffSupport()` are in a separate loop.